pkg/cache/v2/cache.go

```diff
@@ -40,32 +41,97 @@ type ConfigWatcher interface {
 	//
 	// Cancel is an optional function to release resources in the producer. If
 	// provided, the consumer may call this function multiple times.
-	CreateWatch(Request) (value chan Response, cancel func())
+	CreateWatch(Request) (value chan ResponseIface, cancel func())
 }
 
 // Cache is a generic config cache with a watcher.
 type Cache interface {
 	ConfigWatcher
 
 	// Fetch implements the polling method of the config cache using a non-empty request.
-	Fetch(context.Context, Request) (*Response, error)
+	Fetch(context.Context, Request) (ResponseIface, error)
+}
+
+type ResponseIface interface {
+	GetDiscoveryResponse() (*discovery.DiscoveryResponse, error)
+
+	GetRequest() *discovery.DiscoveryRequest
+
+	GetVersion() string
 }
 
 // Response is a pre-serialized xDS response.
 type Response struct {
+	ResponseIface
 	// Request is the original request.
 	Request discovery.DiscoveryRequest
 
 	// Version of the resources as tracked by the cache for the given type.
 	// Proxy responds with this version as an acknowledgement.
 	Version string
 
-	// The value indicating whether the resource is marshaled, and only one of `Resources` and `MarshaledResources` is available.
-	ResourceMarshaled bool
-
 	// Resources to be included in the response.
 	Resources []types.Resource
 
+	// The value indicating whether the resource is marshaled, and only one of `Resources` and `MarshaledResources` is available.
+	isResourceMarshaled bool
+
 	// Marshaled Resources to be included in the response.
-	MarshaledResources []types.MarshaledResource
+	marshaledResponse *discovery.DiscoveryResponse
+}
+
+type PassthroughResponse struct {
+	ResponseIface
+	// Request is the original request.
+	Request discovery.DiscoveryRequest
+
+	// Resources to be included in the response.
+	Response *discovery.DiscoveryResponse
+}
+
+func (r Response) GetDiscoveryResponse() (*discovery.DiscoveryResponse, error) {
+	if r.isResourceMarshaled {
+		return r.marshaledResponse, nil
+	}
+
+	marshaledResources := make([]*any.Any, len(r.Resources))
```

Comment:
Might be worth adding a comment here to mention we marshal once so that subsequent calls don't require re-marshaling.